### PR TITLE
♻️(backend) refactor post hackathon to a first working version

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

This project was copied from https://github.com/numerique-gouv/people and hacked to make a POC in a 2-day hackathon. We need to clean and refactor things in order to get a first version of the product we want.

## Proposal

Make backend work as expected by our proof of concept:
- remove unrelated code
- refactor until tests pass

### CODE REVIEW

File: src/backend/core/tests/test_api_template_accesses.py [hunk 1]
```diff
@@ -0,0 +1,824 @@
...
+    template = factories.TemplateFactory(users=[(user, "administrator")])
+    access = factories.TemplateAccessFactory(
+        template=template,
+        role=random.choice(["administrator", "member"]),
+    )
+    old_values = serializers.TemplateAccessSerializer(instance=access).data
+
+    new_values = {
+        "id": uuid4(),
+        "user_id": factories.UserFactory().id,
```
Comment (lebaudantoine) [inline src/backend/core/tests/test_api_template_accesses.py:449]: can we update user?

File: src/backend/core/tests/test_api_template_accesses.py [hunk 2]
```diff
@@ -0,0 +1,824 @@
...
+
+        access.refresh_from_db()
+        updated_values = serializers.TemplateAccessSerializer(instance=access).data
+        assert updated_values == old_values
+
+
+def test_api_template_accesses_update_owner_except_owner():
+    """
+    A user who is an owner in a template should be allowed to update
+    a user access for this template except for existing "owner" accesses.
```
Comment (lebaudantoine) [inline src/backend/core/tests/test_api_template_accesses.py:560]: Why an owner cannot update other owner?

**Review by lebaudantoine — APPROVED** [review]
LGTM

File: src/backend/core/tests/test_models_teams.py [hunk 3]
```diff
@@ -1,135 +0,0 @@
-"""
```
[the remark this answers is not in the document]
Comment (lebaudantoine) [reply src/backend/core/tests/test_models_teams.py:1]: I guess, the tests were outdated

File: src/backend/core/models.py [hunk 4]
```diff
@@ -266,14 +220,15 @@ def get_abilities(self, user):
...
-            "get": True,
-            "patch": is_owner_or_admin,
-            "put": is_owner_or_admin,
-            "delete": role == RoleChoices.OWNER,
+            "destroy": role == RoleChoices.OWNER,
+            "generate_document": can_get,
             "manage_accesses": is_owner_or_admin,
+            "update": is_owner_or_admin,
+            "retrieve": can_get,
         }
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/models.py:234]: No and in the next PR I will transform the team from a foreign key to a string and add a DB constraint to ensure that either the user or the team is set but not both at the same time.

File: src/backend/core/factories.py [hunk 5]
```diff
@@ -19,48 +19,39 @@ class UserFactory(factory.django.DjangoModelFactory):
...
-        django_get_or_create = ("sub",)
-
-    user = factory.SubFactory(UserFactory)
     sub = factory.Sequence(lambda n: f"user{n!s}")
     email = factory.Faker("email")
+    language = factory.fuzzy.FuzzyChoice([lang[0] for lang in settings.LANGUAGES])
+    password = make_password("password")
 
 
-class TeamFactory(factory.django.DjangoModelFactory):
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/factories.py:38]: In the next PR, the team becomes just a string in the TeamAccess model.

File: src/backend/core/tests/test_models_teams.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/tests/test_models_teams.py:1]: The Team model will be removed in the next PR.

File: src/backend/core/tests/templates/test_api_templates_create.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    )
+
+    assert response.status_code == 401
+    assert not Template.objects.exists()
+
+
+def test_api_templates_create_authenticated():
+    """
+    Authenticated users should be able to create templates and should automatically be declared
+    as the owner of the newly created template.
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/tests/templates/test_api_templates_create.py:30]: I would say yes. What other condition could be want them to fulfill?

File: src/backend/core/tests/templates/test_api_templates_list.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    assert content["next"] is None
+    assert content["previous"] == "http://testserver/api/v1.0/templates/"
+
+    assert len(content["results"]) == 1
+    template_ids.remove(content["results"][0]["id"])
+    assert template_ids == []
+
+
+def test_api_templates_list_authenticated_distinct():
+    """A template with several related users should only be listed once."""
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/tests/templates/test_api_templates_list.py:109]: Probably had the problem of a duplicated item in the list and trying to prove the efficiency of a distinct clause?

File: src/backend/core/api/permissions.py [hunk 8]
```diff
@@ -45,10 +53,10 @@ def has_object_permission(self, request, view, obj):
...
 
-class AccessPermission(IsAuthenticated):
+class AccessPermission(permissions.BasePermission):
     """Permission class for access objects."""
 
     def has_object_permission(self, request, view, obj):
         """Check permission for a given object."""
         abilities = obj.get_abilities(request.user)
-        return abilities.get(request.method.lower(), False)
+        return abilities.get(view.action, False)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/api/permissions.py:62]: we use action in place of method names as abilities items.

File: src/backend/core/models.py [hunk 9]
```diff
@@ -120,80 +133,20 @@ class Meta:
...
-            ),
-        ]
-
     def __str__(self):
-        main_str = "[main]" if self.is_main else ""
-        return f"{self.email:s}{main_str:s}"
-
-    def clean(self):
-        """Normalize the email field and clean the 'is_main' field."""
-        self.email = User.objects.normalize_email(self.email)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/models.py:176]: As discussed IRL, we decided to trust the OIDC token.

File: src/backend/core/models.py [hunk 10]
```diff
@@ -77,7 +71,26 @@ def save(self, *args, **kwargs):
...
+        help_text=_(
+            "Required. 255 characters or fewer. Letters, numbers, and @/./+/-/_ characters only."
+        ),
+        max_length=255,
+        unique=True,
+        validators=[sub_validator],
+        blank=True,
+        null=True,
+    )
+    email = models.EmailField(_("email address"), unique=True, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/models.py:93]: After checking, we set on the model:
```python
USERNAME_FIELD = "email"
```
So it has to be unique. The only solution would be to have 2 separate fields: one for the identity and one for the admin login :thinking:

[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/models.py:93]: So I did that as it seems to be a good idea to be as permissive as possible on the email transmitted by the OIDC token which we trust...

File: src/backend/core/tests/test_models_template_accesses.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    """Check abilities of owner access for the owner of a template."""
+    access = factories.TemplateAccessFactory(role="owner")
+    factories.TemplateAccessFactory(template=access.template)  # another one
+    user = factories.TemplateAccessFactory(template=access.template, role="owner").user
+    abilities = access.get_abilities(user)
+    assert abilities == {
+        "destroy": False,
+        "retrieve": True,
+        "update": False,
+        "set_role_to": [],
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/backend/core/tests/test_models_template_accesses.py]: Agreed, let's allow this.

File: src/backend/core/tests/test_api_template_accesses.py (hunk 2 above)
Comment (sampaccoud) [reply src/backend/core/tests/test_api_template_accesses.py]: Agreed, let's allow this.

**Comment by sampaccoud** [comment]
@lebaudantoine maybe you should checkout the [fixup commit](https://github.com/numerique-gouv/impress/pull/4/commits/c1e2b63e9f73442d89af01f1b582941d9b3c2a89) before merge.

**Review by lebaudantoine — COMMENTED** [review]
LGTM, minor comments in the admin?
Capture d’écran 2024-02-23 à 13 52 05

**Comment by sampaccoud** [comment]
@lebaudantoine all good points :pray: I improved the whole user admin view to match what was done on https://github.com/numerique-gouv/people.

**Pull request merged**